### ISSUE TITLE
(core): unused cross-region exports can't be deleted

### ISSUE THREAD

**Issue by shahin** [aws/aws-cdk#27902]
### Describe the bug

If a stack maintains multiple cross-region exports, and some subset of cross-region exports are being imported by other stacks, attempting to remove any cross-region export value that is not being used (imported) by any deployed stack:

```
❯ cdk diff [...] --exclusively
[...]
[~] Custom::CrossRegionExportWriter ExportsWriteruswest2[...]
ExportsWriteruswest2[...]
 └─ [~] WriterProps
     └─ [~] .exports:
         └─ [-] Removed: ./cdk/exports/unused-value/anotherstackuseast1Ref[...]certificate[...]
```

Fails with the following error on `cdk deploy`:

```
ExportsWriteruswest2[...]/Resource/Default (ExportsWriteruswest2[...]) Received response status [FAILED] from custom resource. Message returned: Error: Exports cannot be updated: 

    at throwIfAnyInUse (/var/task/index.js:4:10)
```

Other issues don't describe this bug:
- several closed issues reference cross-region exports (https://github.com/aws/aws-cdk/issues/24464, https://github.com/aws/aws-cdk/issues/25114), but these concern creating exports not deleting them
- many closed issues reference removing cross-stack exports (https://github.com/aws/aws-cdk/issues/7602), but cross-stack exports that are not also cross-region don't use the ExportsWriter lambda and can't produce the failure mode
- the [Stack.export_value](https://docs.aws.amazon.com/cdk/api/v2/python/aws_cdk/Stack.html#aws_cdk.Stack.export_value) function does not export values across regions, only across stacks, so it can't be used to create a "dummy" cross-region export for this ExportsWriter lambda

### Expected Behavior

CDK should fail to deploy with the error above only if the `cdk diff` shows an export being removed that is currently imported by another stack.

### Current Behavior

CDK fails to deploy with the error above even when `cdk diff` shows only exports being removed that are *not* currently imported by another stack.

### Reproduction Steps

1. Create two cross-region exports, exported from stack A into another stack B in a different region.
2. Remove one of the two imports from stack B and deploy stack B exclusively. Observe that one import is removed from stack B's `["CrossRegionExportReader"]["ReaderProps"]` template in the CloudFormation console.
3. Attempt to deploy stack A. Observe one of the two exports being removed in the diff. Observe that on the deploy, the ExportWriter fails to update and the deploy fails with the error above.

### CDK CLI Version

2.94.0 (build 987c329)

### Node.js Version

v18.17.1

### OS

macOS

### Language

Python

### Language Version

Python (3.10.4)

**Comment by khushail** [comment]
Hi @shahin , could you please share the sample code so that this issue can be reproduced

**Comment by shahin** [comment]
Thanks for your attention to this issue. The reproduction steps above are the best I can do for now.

**Comment by renschler** [comment]
I am also facing this issue

**Comment by renschler** [comment]
I am facing this issue as well. @shahin do you know if there is a workaround? Is manually deleting the SSM parameters before deploy fix things? Im debuggin in a dev environment now but trying to figure out how I'm going to roll my changes out to prod without a crash; makes me nervous haha

**Comment by shahin** [comment]
@renschler sorry, I don't know of a workaround in CDK. Deleting the SSM parameters might be part of the solution but did not by itself unblock deployment for us.

> (Note there is no way for me to deploy stack B exclusively from what I can tell because I've specified that stack B is dependent on stack A).

If stack B depends on stack A, you can still deploy stack B by itself with the `--exclusively` flag:
```
❯ cdk deploy --help
cdk deploy [STACKS..]

[...]
-e, --exclusively          Only deploy requested stacks, don't include
                             dependencies                              [boolean]
```

**Comment by renschler** [comment]
thanks @shahin 

@khushail I made a repoduction repo please let me know if you have any questions

https://github.com/renschler/cross-region-delete

**Comment by renschler** [comment]
ps as a warning to others, don't delete those unnecessary SSM paramters, it will create more issues! I'm now having to re-add them manually to SSM in every region where I deleted them.

**Comment by NickDarvey** [comment]
Given the bug found by @elliotsegler in https://github.com/aws/aws-cdk/issues/29699, it looks like the error `Error: Exports cannot be updated` is explicitly thrown when an SSM parameter has the tag `aws-cdk:strong-ref:SomeStackName` [^1].

Perhaps a workaround is removing that tag and then redeploying? I understand the tag is there to stop updates [^2], but I'm not sure why that's necessary [^3]. (So this workaround might not be a very good workaround!) In my case I am making a change that removes a cross-region reference so... should be fine?

[^1]: https://github.com/aws/aws-cdk/blob/36fd79d8714bd29527bb1184ec10cd504b83510d/packages/%40aws-cdk/custom-resource-handlers/lib/core/cross-region-ssm-writer-handler/index.ts#L92-L137
[^2]: https://github.com/aws/aws-cdk/blob/b82320b08ebcda98b85be8ceb56a5a4b39511d4a/packages/aws-cdk-lib/core/adr/cross-region-stack-references.md?plain=1#L243-L247
[^3]: My best bet is that CloudFormation won't detect a change in the parameter's value for the importing stack, so resources using the parameter won't actually get updated with the new value. The `strong-ref` check is there to stop people getting confused about why their resourced don't get updated with the new value.

**Comment by elliotsegler** [comment]
@NickDarvey I'm wondering if that's an ordering issue...

The cross-region-reader-handler and associated `Custom::CrossRegionReader` resource get added to the stack that consumes the exports. That handler is capable of removing the tags[^1].

In my issues I was letting CDK resolve the order of operations, and it was attempting to delete resources in the stack that contained the exports _before_ updating the reader stack which would have removed those tags. I worked around it by specifically calling a deployment of the reader stack, then a second phase of deployments to the stack that contained now unused exports.

[^1]: https://github.com/aws/aws-cdk/blob/36fd79d8714bd29527bb1184ec10cd504b83510d/packages/%40aws-cdk/custom-resource-handlers/lib/core/cross-region-ssm-reader-handler/index.ts#L31-L33

**Comment by NickDarvey** [comment]
>  I worked around it by specifically calling a deployment of the reader stack

That would have worked for me too actually!

**Comment by andreialecu** [comment]
Looks like we got hit by this too. The stack is now in `UPDATE_ROLLBACK_FAILED` state and cannot seemed to be rolled back. 

Any workarounds to get it un stuck? Would skipping the resource be safe?

**Comment by candu** [comment]
I can see the challenge here: if I export a cross-region reference from stack A to stack B, and stack B depends on resources in stack A, and I delete the resource this cross-region reference refers to, I then have a cyclic update dependency.

From the ADR linked above, I'd want to update stack B (export reader), then stack A (export writer).

However, since stack B depends on resources in stack A, I'd want to update stack A, then stack B.

As a result, neither stack is safe to update before the other.

Maybe an ideal resolution would be to detect the cyclic dependency here and fast-fail?  In cases like this, I'd much rather have fast-fail behaviour than unblock `UPDATE_ROLLBACK_FAILED` snags (e.g. with a warning that guides me to use `--exclusively`, or to break up the update into two steps).

**Comment by AposLaz** [comment]
Any updates for this issue?

**Comment by ruslan-roboto** [comment]
This is giving us headaches as well. Any update would be immensely helpful. Thanks!

**Comment by flochaz** [comment]
On my side, got it fixed by:
1. Reverted manually to old values (check your CustomCrossRegionExportWriter Lambda logs in cloudwatch to get the old and new value and be careful in the logs between the one that corespond to the initial rewrite and the one triggered by the rollback :) ) **WITHOUT THE aws-cdk:strong-ref TAG !**
1. Trigger a "Continue update rollback" of the cloudformation stack in UPDATE_ROLLBACK_FAILED without the failing CustomCrossRegionExportWriter resource (check the box in the advanced continue rollback modal)
1. Redeploy the stacks

**Comment by JonWallsten** [comment]
I see two EASY solutions conceptually, but maybe harder technically.
Both would require to allow a way to keep the reference in the ExportWriter resource, but remove it from the ExportReader. 
a) Either by adding it manually to the ExportWriter in the producing stack
b) Have some kind of DummyImport in the receiving stack that only creates the export and not the import

Then you can really easily create a new resource along the old if you need to replace something, like a Certificate, or remove the resource if that's what you need.

**Comment by renschler** [comment]
Coming back to this issue because I hit it again.

For context:
- I have a lambda function that triggers statemachines in other regions. There's currently an issue where if your crossRegionReferences parameters are too long the deploy fails https://github.com/aws/aws-cdk/issues/30119 .
- I ran into that bug, to fix things I thought I could just remove the additional regions and try the deploy again.
- This failed with the error ```Exports cannot be updated``` and the UPDATE_ROLLBACK_FAILED

[will update this comment later when I remember how I sorted this out last time]

**Comment by renschler** [comment]
Found some more references to this issue:
https://x.com/theburningmonk/status/1826268576187060613
https://sst.dev/blog/moving-away-from-cdk/#3-export-in-use

**Comment by horyd** [comment]
I managed to fix my cross region issues, finally. For context, I had a certificate in us-east-1 that I wanted to make changes to (or more realistically, recreate and remove the old one). The certificate was used in a cloudfront distribution in my second stack in eu-west-2.

1. Create a new certificate in stack 1. In stack 2, reference the new certificate in the CF distribution. Also in stack 2, add a CFNOutput for the original certificate (in my case I output the original certificate ARN). This second step is critical, as it makes sure the ExportsWriter does not get removed in stack 1 (which is what causes failures).
2. Deploy both stacks.
3. Remove the CFNOutput from stack 2. **EXCLUSIVELY** deploy stack 2 using --exclusively. This will remove the reader.
4. Deploy stack 1. It will see the reference to the old certificate is no longer needed and remove the writer.
5. Remove the certificate and deploy stack 1 again.
